flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointsCleaner.java

```diff
@@ -71,10 +81,19 @@ public void cleanCheckpoint(
             boolean shouldDiscard,
             Runnable postCleanAction,
             Executor executor) {
-        Checkpoint.DiscardObject discardObject =
-                shouldDiscard ? checkpoint.markAsDiscarded() : Checkpoint.NOOP_DISCARD_OBJECT;
-
-        cleanup(checkpoint, discardObject::discard, postCleanAction, executor);
+        LOG.debug(
+                "Clean checkpoint {} parallel-mode={} shouldDiscard={}",
+                checkpoint.getCheckpointID(),
+                parallelMode,
+                shouldDiscard);
+        if (parallelMode) {
+            cleanupParallel(checkpoint, shouldDiscard, postCleanAction, executor);
+        } else {
+            Checkpoint.DiscardObject discardObject =
+                    shouldDiscard ? checkpoint.markAsDiscarded() : Checkpoint.NOOP_DISCARD_OBJECT;
+
+            cleanup(checkpoint, discardObject::discard, postCleanAction, executor);
```

Comment:
nit: maybe remove this method, and replace it with a call:
```
cleanupParallel(checkpoint, shouldDiscard, postCleanAction, Executors#newDirectExecutorService);
```
? Wouldn't this behave identically?

Reply:
looks like DefaultCompletedCheckpointStore#shutdown calls the cleanCheckpoint from main thread instead of the ioExecutor. Combined into one method.